registry/virtual.go

```diff
@@ -1,50 +1,455 @@
 	db, err := getDBForVirtualSpace(virtualSpaceName)
 	if err != nil {
-		return nil
+		return err
+	}
+
+	virtualSpace, ok := base.Config.VirtualSpaces[virtualSpaceName]
+	if !ok {
+		return fmt.Errorf("unable to find virtual space %s", virtualSpaceName)
+	}
+
+	spaceName := virtualSpace.Source
+	s, ok := space.GetSpace(spaceName)
+	if !ok {
+		return fmt.Errorf("unable to find %s space", spaceName)
 	}
 
 	overwrite, err := findOverwrite(db, appSlug)
 	if err != nil {
-		return nil
+		return err
 	}
 
-	shasum, ok := overwrite["icon"].(string)
-	if !ok || shasum == "" {
-		return nil
+	var regenerated []*Version
+
+	for _, channel := range Channels {
+		lastVersion, err := FindLatestVersion(s, appSlug, channel)
+		if err != nil {
+			return err
+		}
+		if lastVersion == nil || versionOverwrittenAlreadyProcessed(regenerated, lastVersion) {
+			continue
+		}
+
+		tarball, err := getOriginalTarball(s, lastVersion)
+		if err != nil {
+			return err
+		}
+		file, manifest, icon, err := generateOverwrittenTarball(lastVersion, overwrite, tarball)
+		// Tricky return, last part
+		// Even in case of error, we must to take care of any file returned to be sure to erase it from the fs
+		if file != "" {
+			defer func() {
+				cerr := os.Remove(file)
+				if err == nil {
+					err = cerr
+				}
+			}()
+		}
+		if err != nil {
+			return err
+		}
+
+		hash, size, err := storeOverwrittenTarball(virtualSpace, file)
+		if err != nil {
+			return err
+		}
+
+		newVersion := lastVersion.Clone()
+		newVersion.Rev = ""
+		newVersion.AttachmentReferences = map[string]string{"tarball": hash}
+		newVersion.Size = size
+		newVersion.Sha256 = hash
+
+		u, err := url.Parse(newVersion.URL)
+		if err != nil {
+			return err
+		}
+		u.Path = path.Join(virtualSpace.Name, u.Path)
+		newVersion.URL = u.String()
+
+		if icon != "" {
+			newVersion.AttachmentReferences["icon"] = icon
+		}
+		j, err := json.Marshal(manifest)
+		if err != nil {
+			return err
+		}
+		newVersion.Manifest = j
+
+		existingVersion, err := findOverwrittenVersion(virtualSpace, lastVersion)
+		if err != nil {
+			return err
+		}
+		if existingVersion != nil {
+			newVersion.Rev = existingVersion.Rev
+			// We already have a version, destroy the old tarball if changed
+			h, ok := existingVersion.AttachmentReferences["tarball"]
+			if ok && hash != h {
+				prefix := base.Prefix(virtualSpace.Name)
+				if err := base.Storage.Remove(prefix, h); err != nil {
+					return err
+				}
+			}
+		}
+
+		db := virtualSpace.VersionDB()
+		if _, err = db.Put(context.Background(), newVersion.ID, newVersion); err != nil {
+			return err
+		}
+
+		regenerated = append(regenerated, lastVersion)
 	}
 
-	content, headers, err := base.GlobalAssetStore.Get(shasum)
+	return nil
+}
+
+// FindAppOverride finds if the app have overwritten value in the virtual space
+func FindAppOverride(virtualSpace *base.VirtualSpace, appSlug string, name string) (*string, error) {
```

Comment:
Why do you return a pointer to a string, not just a string?

Reply:
Yeah, we can return "" if not found.